lookout/style/typos/analyzer.py

```diff
@@ -145,17 +151,114 @@ def render_comment_text(self, typo_fix: TypoFix) -> str:
+        suggestions, confidences = [], []
+        for final_sugg, conf in self.generate_identifier_suggestions(
+                typo_fix.candidates, typo_fix.identifier, n_candidates=self.config["n_candidates"],
+        ):
+            suggestions.append(self.reconstruct_identifier(self.parser, final_sugg,
+                                                           typo_fix.identifier))
+            confidences.append(conf)
+        return self.comment_template.render(
+            identifier=typo_fix.identifier,
+            suggestions=suggestions,
+            old_code_line=file_lines[typo_fix.line_number - 1],
+            confidences=self._normalize_confidences(confidences),
+            zip=zip,
+            enumerate=enumerate,
+            int=int,
+        )
+
+    @staticmethod
+    def _compute_confidence(confidences: Sequence[float]) -> float:
+        confidence = 1
+        for conf in confidences:
+            confidence *= conf
+        return confidence
 
     @staticmethod
-    def _get_comment_confidence(suggestions: Sequence[Candidate]) -> int:
-        invert_confidence = 1
-        for suggestion in suggestions:
-            invert_confidence *= (1 - suggestion.confidence)
-        return int(100 * (1 - invert_confidence))
+    def _normalize_confidences(confidences: Sequence[float]):
+        s = sum(confidences)
+        return [conf / s for conf in confidences]
+
+    def generate_identifier_suggestions(self, suggestions: Mapping[str, Iterable[Candidate]],
+                                        identifier: str, n_candidates: int = 3) \
+            -> Iterator[Tuple[List[str], float]]:
+        """
+        Generate suggestions for the identifier and compute the probability of suggestion.
+
+        Example: `same_wrang_vor`
+        1) split identifier: `same_wrang_vor` -> initial_tokens [`same`, `wrang`, `vor`]
+        2) candidates
+        candidates\intital token|`same`     |`wrang`     |`vor`
+        candidate 1             |`same`, 0.9|`range`, 0.8|`for`, 0.8
+        candidate 2             |`some`, 0.8|`wrong`, 0.7|`var`, 0.6
+        candidate 3             |           |`lang`, 0.5 |
+        3) select best suggestion per token
+        candidates\intital token|`same`     |`wrang`     |`vor`
+        candidate 1             |`same`, 0.9|`range`, 0.8|`for`, 0.8
+        and it will give us:
+        `final_sugg` :  [`same`, `range`, `for`]
+        `confidences`: [0.9, 0.8, 0.8]
+        4) put other candidates in a queue where the first element in tuple is position in
+           `final_sugg`
+        [(0, `some`, 0.8), (1, `wrong`, 0.7), (2, `var`, 0.6), (1, `lang`, 0.5)]
+        and sort queue by probability
+        5) generate next identifier suggestions using sorted candidates -
+        ------------------------------------
+            next candidate: (0, `some`, 0.8)
+            `final_sugg` :  [`some`, `range`, `for`]
```

Comment:
I don't understand why you keep modifying `final_sugg` with every new candidate? So in the end all additional candidates have less probable `"some"` instead of `"same"`, which seems definitely wrong to me. Imo `final_sugg` should be fixed in the form you returned first, and then replace one token just for one yield,  while leaving `final_sugg`  as it is as a standard.

Reply:
good point. In this case @zurk suggestion about `itertools.product` should be used.